### PR TITLE
[diem-vm] Implement a prefetching logic for diem-vm

### PR DESCRIPTION
**Pull request by runtian-zhou**
## Motivation

This is the revised version of #7338. The logic here is to prefetch the account blobs into the cache of the state view so that we can have account blob deserialization done in background.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/diem/diem/blob/main/CONTRIBUTING.md#pull-requests)?

Yes
## Test Plan

Copied from #7338:

1. Launching one thread to prefetch accounts:
```
bench_p2p_small         time:   [410.93 ms 416.28 ms 421.76 ms]
                        change: [-6.6657% -4.3640% -2.2150%] (p = 0.00 < 0.05)
                        Performance has improved.

bench_p2p_medium        time:   [769.03 ms 775.82 ms 783.25 ms]
                        change: [-4.9727% -3.8442% -2.7737%] (p = 0.00 < 0.05)
                        Performance has improved.

bench_p2p_large         time:   [771.35 ms 786.97 ms 803.10 ms]
                        change: [-6.2914% -2.7613% +0.2655%] (p = 0.14 > 0.05)
                        No change in performance detected.
```
I've been observing a constant perf increase on medium case with 1000 txns in a block. The perf number for small and large has been pretty fluctuated but mostly on the improvement side.

2. Launching multiple threads to prefetch accounts:
```
bench_p2p_small         time:   [408.61 ms 411.82 ms 414.91 ms]
                        change: [-7.4353% -5.3868% -3.6983%] (p = 0.00 < 0.05)
                        Performance has improved.

bench_p2p_medium        time:   [761.75 ms 772.99 ms 783.74 ms]
                        change: [-5.6370% -4.1952% -2.7912%] (p = 0.00 < 0.05)
                        Performance has improved.

bench_p2p_large         time:   [736.94 ms 744.87 ms 753.19 ms]
                        change: [-10.851% -7.9634% -5.7760%] (p = 0.00 < 0.05)
                        Performance has improved.
```
This version is performing constantly better then the first version, especially on the p2p_large case. I would assume it's because we have more threads prefetching the values.

With this benchmark result, I would suggest land the version with multiple prefetching threads and fine tune the number of threads in the future.

### CODE REVIEW

File: language/diem-vm/src/diem_transaction_executor.rs [hunk 1]
```diff
@@ -621,6 +622,39 @@ impl DiemVM {
         ))
     }
 
+    fn preload_cache(
+        signature_verified_block: &[PreprocessedTransaction],
+        data_view: &dyn StateView,
```
Comment (vgao1996) [inline language/diem-vm/src/diem_transaction_executor.rs:627]: `&impl StateView` instead of `&dyn StateView`?

File: storage/storage-interface/src/state_view.rs [hunk 2]
```diff
@@ -168,11 +168,16 @@ impl<'a> StateView for VerifiedStateView<'a> {
                             err
                         )
                     })?;
-                assert!(self
-                    .account_to_proof_cache
+                self.account_to_proof_cache
                     .write()
-                    .insert(address_hash, proof)
-                    .is_none());
+                    .insert(address_hash, proof);
```
Comment (bob-wilson) [inline storage/storage-interface/src/state_view.rs:173]: I don't see much point in keeping this assertion as a comment. Can you move the comment about multiple threads before the `self.account_to_proof_cache` line and rephrase it to explain why the `insert()` result may not be `none`?

**Review by bob-wilson — APPROVED** [review]
